File: 986. Interval List Intersections.py
class Solution:
    def intervalIntersection(self, A: List[List[int]], B: List[List[int]]) -> List[List[int]]:
        result = []
        pa = 0
        pb = 0
        while pa < len(A) and pb < len(B):
            if A[pa][0] > B[pb][1]:
                pb += 1
            elif A[pa][1] < B[pb][0]:
                pa += 1
            else:
                inters = [max(A[pa][0], B[pb][0]), min(A[pa][1], B[pb][1])]
                result.append(inters)
                if A[pa][1] < B[pb][1]:
                    pa += 1
                else:
                    pb += 1
        return result

File: 986. Interval List Intersections.java
class Solution {
    public int[][] intervalIntersection(int[][] A, int[][] B) {
        if (A==null || A.length==0 || B==null || B.length==0) {
            return new int[0][2];
        }
        int pa=0, pb=0;
        int[] a, b;
        List<int[]> ls = new ArrayList<>();
        while (pa < A.length && pb < B.length) {
            a = A[pa];
            b = B[pb];
            if (a[1] < b[0]) {
                pa++;
            } else if (a[0] > b[1]) {
                pb++;
            } else {
                int[] in = new int[2];
                in[0] = Math.max(a[0], b[0]);
                in[1] = Math.min(a[1], b[1]);
                ls.add(in);
                if (a[1] < b[1]) {
                    pa++;
                } else {
                    pb++;
                }
            }
        }
        int[][] ans = new int[ls.size()][2];
        for (int i=0; i<ans.length; i++) {
            ans[i] = ls.get(i);
        }
        return ans;
    }
}